utils/wasm-builder/src/wasm_project.rs

```diff
@@ -261,35 +261,93 @@ extern "C" fn metahash() {{
         Ok(())
     }
 
-    /// Post-processing after the WASM binary has been built.
-    ///
-    /// - Copy WASM binary from `OUT_DIR` to `target/wasm32-unknown-unknown/<profile>`
-    /// - Generate optimized and metadata WASM binaries from the built program
-    /// - Generate `wasm_binary.rs` source file in `OUT_DIR`
-    pub fn postprocess(&self) -> Result<()> {
-        let file_base_name = self
-            .file_base_name
-            .as_ref()
-            .expect("Run `WasmProject::generate()` first");
-
-        let from_path = self
-            .target_dir
-            .join(format!("wasm32-unknown-unknown/{}", self.profile))
-            .join(format!("{}.wasm", &file_base_name));
-
-        fs::create_dir_all(&self.target_dir)?;
-        fs::create_dir_all(&self.wasm_target_dir)?;
+    pub fn postprocess_meta(
+        &self,
+        original_wasm_path: &PathBuf,
+        file_base_name: &String,
+    ) -> Result<()> {
+        let meta_wasm_path = self
+            .wasm_target_dir
+            .join([file_base_name, ".meta.wasm"].concat());
+
+        if smart_fs::check_if_newer(original_wasm_path, &meta_wasm_path)? {
+            fs::write(
+                meta_wasm_path.clone(),
+                Optimizer::new(original_wasm_path.clone())?.optimize(OptType::Meta)?,
+            )?;
+        }
 
-        let [to_path, to_opt_path, to_meta_path] = [".wasm", ".opt.wasm", ".meta.wasm"]
-            .map(|ext| self.wasm_target_dir.join([file_base_name, ext].concat()));
+        smart_fs::write(
+            self.out_dir.join("wasm_binary.rs"),
+            format!(
+                r#"#[allow(unused)]
+                       pub const WASM_BINARY: &[u8] = include_bytes!("{}");
+                       #[allow(unused)]
+                       pub const WASM_EXPORTS: &[&str] = &{:?};"#,
+                display_path(meta_wasm_path.clone()),
+                Self::get_exports(&meta_wasm_path)?,
+            ),
+        )
+        .context("unable to write `wasm_binary.rs`")
+        .map_err(Into::into)
+    }
 
-        // Optimize source.
-        if !self.project_type.is_metawasm()
-            && smart_fs::copy_if_newer(&from_path, &to_path).context("unable to copy WASM file")?
-        {
-            _ = optimize::optimize_wasm(to_path.clone(), to_path.clone(), "s", false);
+    pub fn postprocess_opt(
+        &self,
+        original_wasm_path: &PathBuf,
+        file_base_name: &String,
+    ) -> Result<()> {
+        let opt_wasm_path = self
```

Comment:
Note, now original wasm file and it's copy in `target/wasm32-unknown-unknown/release(debug)/` will never be changed by optimisations, for more clear wasm-builder work.